### ISSUE TITLE
Increase Coverage of Test Suite

### ISSUE THREAD

**Issue by fubuloubu** [yearn/yearn-vaults#156]
Analyze coverage gaps for `Vault`/`BaseStrategy`/`Registry` (using `brownie gui` after `brownie test tests/functional --coverage`) and look for ways to increase coverage of calling conditions the test suite is currently missing.

Examples:
- [x] `token.transfer()/transferFrom()` return False
- [x] `Vault.report()` w/ 0% fees & profit
- [x] `Vault.report()` w/ profit/debt payment but no tokens
- [x] `token.transfer()` w/ no return value
- [x] `Strategy.harvestTrigger()` w/ debt outstanding
- [x] `Strategy.harvestTrigger()` w/ realized loss
- [x] `Strategy.sweep()` w/ protected tokens
- [x] `Strategy.setKeeper()/setReward()/setMinReportDelay()/setProfitFactor()/setDebtThreshold()` w/ governance

**Comment by steffenix** [comment]
Would like to help!

**Comment by steffenix** [comment]
will help

**Comment by steffenix** [comment]
I have looked at the reports for `Vault.report()` I do not see lines highlighted in orange on `_assessFees` could you elaborate the examples?

**Comment by fubuloubu** [comment]
![image](https://user-images.githubusercontent.com/3859395/105633910-548b0880-5e29-11eb-81aa-f22500f34b02.png)

**Comment by steffenix** [comment]
So it's about testing the leftover dust?

**Comment by fubuloubu** [comment]
No, that scenario shoukd be skipped if you set the fees for governance to 0%

**Comment by steffenix** [comment]
I guess you know the tests enough so that you know that a governance fee of 0 percent isn't tested because I can't understand that looking at the test coverage.

**Comment by fubuloubu** [comment]
> I guess you know the tests enough so that you know that a governance fee of 0 percent isn't tested because I can't understand that looking at the test coverage.

Haha, yes. Sorry, this is the case.

**Comment by fubuloubu** [comment]
Make sure you rebase your work on https://github.com/iearn-finance/yearn-vaults/pull/164 (once merged)

I noticed one of our test cases wasn't running because it was inadvertently marked as a fixture

**Comment by steffenix** [comment]
What does the next one means?

`Vault.report()` w/ profit/debt payment but no tokens

**Comment by fubuloubu** [comment]
> * `Vault.report()` w/ profit/debt payment but no tokens

It looks like this one actually was resolved

> * `Strategy.harvestTrigger()` w/ debt outstanding
> * `Strategy.harvestTrigger()` w/ realized loss

The only one missing here is:
![image](https://user-images.githubusercontent.com/3859395/106019405-9e295c80-6090-11eb-8e78-104b489ef8f6.png)

**Comment by fubuloubu** [comment]
Also this bit in `BaseStrategy.harvest()`
![image](https://user-images.githubusercontent.com/3859395/106019703-f19baa80-6090-11eb-9c04-8db4010d18d9.png)

**Comment by snorkr4** [comment]
Mind if I help too? I'm new here, figured this would be a great way to start walking through the code.

**Comment by fubuloubu** [comment]
@snorkr4 sure! please coordinate with @steffenix, as he has a few PRs in prog

**Comment by snorkr4** [comment]
@steffenix what's the best way to get in touch? I don't have an invite to the developer discord, if that's where you are. 

I started looking into coverage of Registry. I think I have it up to 100%. I can hold off on creating a PR until I confirm that I'm not going to collide with your changes.

![image](https://user-images.githubusercontent.com/78279638/106371410-dfcc3880-6318-11eb-8c5f-725eb7d67d35.png)

**Comment by snorkr4** [comment]
@steffenix I created a draft PR for now (#177). Assuming it's not colliding with your work, I can put it up for review.

**Comment by steffenix** [comment]
> @steffenix I created a draft PR for now (#177). Assuming it's not colliding with your work, I can put it up for review.

Go ahead I have put this on pause while I was working on other tasks

**Comment by snorkr4** [comment]
> Go ahead I have put this on pause while I was working on other tasks.

Sounds good. I put #177 up for review. @fubuloubu is there documentation of the review process somewhere? E.g. do I assign you as a reviewer, since you filed the issue?

**Comment by snorkr4** [comment]
Thanks for your help, @fubuloubu -- I learned a lot with #177. Looking at the Vault test coverage now.

**Comment by fubuloubu** [comment]
Excellent work getting that to 100%!

**Comment by fubuloubu** [comment]
p.s. @snorkr4 hitting `BaseStrategy`/`TestStrategy` might be easier next step!

**Comment by snorkr4** [comment]
> hitting BaseStrategy/TestStrategy might be easier next step!

Ok sounds good :)

**Comment by snorkr4** [comment]
> `Strategy.sweep()` w/ protected tokens

Should I add protected tokens to `TestStrategy`? I can't think of any way to test sweep w/ protected tokens without modifying `TestStrategy.sol`. I'll assume that's the right approach and get started. Just checking.

**Comment by fubuloubu** [comment]
> > `Strategy.sweep()` w/ protected tokens
> 
> Should I add protected tokens to `TestStrategy`? I can't think of any way to test sweep w/ protected tokens without modifying `TestStrategy.sol`. I'll assume that's the right approach and get started. Just checking.

yes, you are correct. just add a function that allows you push a random token into protected status

**Comment by snorkr4** [comment]
Working on increasing Vault test coverage (WIP #200).

**Comment by fubuloubu** [comment]
@snorkr4 just post snapshots of the coverage diffs from CI w/ highlights for us to review when you are ready! Thanks!

**Comment by snorkr4** [comment]
@fubuloubu I'd say #199 is ready to review. I've posted "before and after" coverage summaries from CI -- not sure how to get a diff from the CI.

A few issues remain for `*Strategy` coverage:

1. For some reason these branches in the authority modifiers show up as not executed. I'm having trouble understanding this.

<img src="https://user-images.githubusercontent.com/78279638/107904664-4e0c2000-6f01-11eb-9aff-433cae78fc4a.png" width=500>

2. `BaseStrategy.sweep` shows up as 87.5% covered in the summary, but I'm not seeing any highlights in the gui.

Branches:
<img src="https://user-images.githubusercontent.com/78279638/107904804-b2c77a80-6f01-11eb-920d-ea8501a1b992.png" width=500>
Statements:
<img src="https://user-images.githubusercontent.com/78279638/107904858-e86c6380-6f01-11eb-903a-3123c2e379df.png" width=500>

3. Similarly, `BaseStrategy.harvest` shows up 80% covered, but no highlights.

Branches:
<img src="https://user-images.githubusercontent.com/78279638/107904957-279ab480-6f02-11eb-9098-0e7e20ea1522.png" width=500>
Statements:
<img src="https://user-images.githubusercontent.com/78279638/107904963-32ede000-6f02-11eb-93d9-23a88f45f702.png" width=500>

4. Same thing with `BaseStrategy.setEmergencyExit` (50% coverage reported, with no highlights).

5. TestStrategy.liquidatePosition has a missing branch:

<img src="https://user-images.githubusercontent.com/78279638/107904584-1a30fa80-6f01-11eb-8d17-8582f3102f13.png" width=500>

Anyway, I'd propose reviewing #199 because it has a good chunk of improvements, and I can work on the above issues in a separate PR.

**Comment by snorkr4** [comment]
To summarize what I'm working on: 

- #199 - improves coverage for BaseStrategy (ready to review).
- #200 is a draft PR with my progress so far on coverage for `Vault` (not ready to review).
- [No PR yet] As mentioned in the previous comment, there will still be some coverage gaps for `*Strategy` -- I plan to work on this after #199 is merged.

**Comment by fubuloubu** [comment]
> For some reason these branches in the authority modifiers show up as not executed. I'm having trouble understanding this.

I think there is a bug with brownie coverage for modifiers. I've been hearing this. Seems like you're running into a few issues here.

---

Yeay, I'll give a look. You didn't run into any weird behaviors or things in the code that didn't work the way you expected?

**Comment by snorkr4** [comment]
> You didn't run into any weird behaviors or things in the code that didn't work the way you expected?

Overall the code has been very readable and has mostly worked out of the box as expected. Some of the accounting in `BaseStrategy` has been hard for me to wrap my head around, so I ran into some challenges there when I was working on `harvest` and `harvestTrigger`. But I think I'm starting to get it, the more cases I work through.

**Comment by fameal** [comment]
Being worked on #226
